### ISSUE TITLE
Control on the weight quantization

### ISSUE THREAD

**Issue by balditommaso** [Xilinx/brevitas#1123]
I am asking this question because I am working with a custom implementation of a `QuantConv2d` layer. During the training, the weights of the layer have to be processed with a series of operations that can be done in full precision, then the quantized version of the processed weights can be used to apply the convolution. 

So far, I have applied the preprocessing on the `. value` version of the quant_weight, breaking the quantization and then I re-apply the quantization making the weights passing through a `QuantIdentity`. However, This approach is sub-optimal because it increases the quantization error and it is tricky to emulate the weight quantization scheme with the activations.

Is there a way to control when the layer apply the weight quantization?

PS: During the inference there are no problems because it will behaive like a traditional `QuantConv2d`.

**Comment by Giuseppe5** [comment]
I would recommend using [parametrize](https://pytorch.org/docs/stable/generated/torch.nn.utils.parametrize.register_parametrization.html).

You register your parametrization, and those will be applied automatically before quantization.

**Comment by balditommaso** [comment]
Sorry one further question: what is the pipeline of quantization during training? Are weights still updated or we should rely only on the `.value` field?

**Comment by Giuseppe5** [comment]
I'm not sure I follow but gradients should be correctly propagated. If you need an example on how to use quant_weights with custom layers, this is the general implementation of a forward pass for a quantized int layer:
https://github.com/Xilinx/brevitas/blob/4617f7bd136e96fa21c7f76e3c7e2e37fe563837/src/brevitas/quant_tensor/int_torch_handler.py#L140

**Comment by Giuseppe5** [comment]
If your layer is a custom QuantConv2d and you end up calling `torch.nn.functional.conv2d` passing a QuantTensor, everything is handled by Brevitas

**Comment by balditommaso** [comment]
My layer extend QuantConv2d, before calling its `inner_forward_impl` I need to do some computations with the weights that must be traced by the gradients, but in the 'inner_forward_impl` I have access only to the quant_weight, however if I move this processing in the `forward` before calling the `inner_forward` and I work with traditional `.weight` does it propagate to the quantization?

**Comment by Giuseppe5** [comment]
Could you post a code snippet to get an idea of what you're trying to achieve?

**Comment by balditommaso** [comment]
Here an example: 
I tried parametrization as you suggested, but there are problems when training on GPU becuase the `scales` are on cpu

```

class Operation(nn.Module):
    @staticmethod
    def forward(weight: Tensor) -> Tensor:
        return weight + torch.ones_like(weight)
    

class CustomQuantConv2d(qnn.QuantConv2d):
    def __init__(self,
            in_channels: int,
            out_channels: int,
            kernel_size: _size_2_t,
            padding: Union[_size_2_t, str] = 'same',
            padding_mode: str = 'circular',
            initializer: Optional[Callable] = torch.nn.init.orthogonal_,
            weight_quant: Optional[WeightQuantType] = Int8WeightPerTensorFloat,
            bias_quant: Optional[BiasQuantType] = None,
            input_quant: Optional[ActQuantType] = None,
            output_quant: Optional[ActQuantType] = None,
            return_quant_tensor: bool = False,
            **kwargs):
        #if padding == 'same':
        #    padding = kernel_size[-1]//2 if isinstance(kernel_size, tuple) else kernel_size//2
        super().__init__(
            in_channels=in_channels,
            out_channels=out_channels,
            kernel_size=kernel_size,
            padding=padding,
            padding_mode=padding_mode,
            weight_quant=weight_quant,
            bias_quant=bias_quant,
            input_quant=input_quant,
            output_quant=output_quant,
            return_quant_tensor=return_quant_tensor,
            **kwargs)
        
        if initializer is not None:
            initializer(self.weight)
            
        # apply rescaling
        register_parametrization(self, "weight", Operation())
    
  
      
```

**Comment by balditommaso** [comment]
This is the error I am getting, I am sure that I am moving correctly all the model on the right GPU because I am training with `pytorch-lightning`.

<img width="817" alt="Screenshot 2024-12-11 alle 11 18 57" src="https://github.com/user-attachments/assets/38e19836-7467-4076-8ab4-7edc3bf98ffb">

**Comment by balditommaso** [comment]
I think the problem is in `_ParameterListStats`, where the `tracked_parameter_list` has the tensors on cpu and then when we move the model on GPU during training they are not moved too.

Hope it is helping!

**Comment by Giuseppe5** [comment]
Might I ask you to pull the latest version of dev?
We made some change to that logic and maybe it will also solve your issue

**Comment by balditommaso** [comment]
With the latest version it is working. Thank you so much!

**Comment by balditommaso** [comment]
Sorry one more question, by using parametrize, the values stored in `quant_weight().value` are the dequantized version of `self.weight` or `self.parametrizations.weight.original`?

I am asking this, becuase I want to be sure that once I export the model, I will get the quantized version of the modified weights.
